File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Map<Integer,Integer> map = new HashMap<>();
    public TreeNode construct(int st_i,int end_i,int st_p, int end_p,int[] postorder,int[] inorder){
        if(st_i > end_i || st_p > end_p)
            return null;
        TreeNode node = new TreeNode(postorder[end_p]);
        int ind = map.get(node.val);
        int left = ind - st_i;
        int right = end_i - ind;
        node.left = construct(st_i,ind - 1 , st_p ,end_p - right - 1,postorder,inorder);
        node.right = construct(ind+1,end_i,end_p - right,end_p - 1,postorder,inorder);
        
        return node;


    }
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        int n = inorder.length;
        for(int i=0;i<n;i++)
            map.put(inorder[i],i);
        return construct(0,n-1,0,n-1,postorder,inorder);


        
    }
}

File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def helper(self,i_l,i_r,p_l ,p_r,inorder,postorder,hashmap):
        if  i_l > i_r or p_l > p_r  :
            return None
        node = TreeNode(postorder[p_r])
        ind = hashmap[node.val]
        l_len = ind - i_l
        r_len = i_r - ind
        node.left = self.helper(i_l,ind - 1,p_l, p_r - r_len - 1 ,inorder,postorder,hashmap)
        node.right = self.helper(ind + 1, i_r , p_r - r_len , p_r - 1 , inorder, postorder, hashmap)
        return node
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        hashmap = {}
        n = len(inorder)
        for i in range(len(inorder)):
            hashmap[inorder[i]] = i
        return self.helper(0 , n - 1 , 0 , n -1 , inorder,postorder,hashmap)





        